### ISSUE TITLE
Checkpointing is slow on nn.DataParallel models

### ISSUE THREAD

**Issue by gpleiss** [pytorch/pytorch#7801]
@wandering007 pointed out this issue in https://github.com/gpleiss/efficient_densenet_pytorch/blob/master/models/densenet.py

I have a model that uses checkpointing on several layers. On a single GPU, the model runs fairly fast (e.g. only a 15-20% overhead). On multiple GPUs, using an `nn.DataParallel` @wandering007 claims that the model runs up to 100x slower.

Here's the important snippets of the model.

```python
def _bn_function_factory(norm, relu, conv):
    def bn_function(*inputs):
        concated_features = torch.cat(inputs, 1)
        bottleneck_output = conv(relu(norm(concated_features)))
        return bottleneck_output
    return bn_function

class _DenseLayer(nn.Module):
    # ...

    def forward(self, *prev_features):
        bn_function = _bn_function_factory(self.norm1, self.relu1, self.conv1)
        bottleneck_output = cp.checkpoint(bn_function, *prev_features)
        # ...
```

There are several `_DenseLayer`s throughout the model.
@wandering007 seems to think that the issue has to do with GPU synchronization? As in, the models must synchronize at every checkpoint during the backward pass.

Original issue is here: https://github.com/gpleiss/efficient_densenet_pytorch/issues/36
Full code of the model is here: https://github.com/gpleiss/efficient_densenet_pytorch/blob/master/models/densenet.py

**Comment by SsnL** [comment]
cc @prigoyal

**Comment by prigoyal** [comment]
Hi @gpleiss, can you provide a _minimalist_ repro of slow speed instead of the entire DenseNet model? 

Also, from the first look, it seems that you are checkpointing every (Conv+Relu+BN) block. This may be inefficient to do to save memory: Relu is inplace, BN output needs to be kept in memory to pass to the next block, input to Conv is saved for doing forward pass again during back propagation. Can you try checkpointing at the level of DenseLayer? This example https://github.com/prigoyal/pytorch_memonger/blob/master/models/optimized/densenet_new.py of densenet checkpointing may be helpful.

**Comment by gpleiss** [comment]
Can do @prigoyal! I'm traveling this week but I'll get you an example next week. And I'll try Checkpoiunting at the DenseLayer level as well.

**Comment by ivankreso** [comment]
I don't understand why does it need to sync between GPUs at every checkpoint? The only reason I can think of is sync between BatchNorm mean and variance but that would cause slowdown without checkpointing also.

@prigoyal DenseNet has (BN+Relu+Conv) units which are then wrapped in np.utils.checkpoint. If I got this right checkpointing feature will then recompute BN and Relu during backward pass but not Conv and only the output of every Conv layer is cached (input to BN is not cached because torch.cat is inside checkpoint function in the code below). This looks optimal to me and my model requires ~ 2x less memory with only 10% slower training time.

```
def _bn_function_factory(norm, relu, conv):
    def bn_function(*inputs):
        concated_features = torch.cat(inputs, 1)
        bottleneck_output = conv(relu(norm(concated_features)))
        return bottleneck_output

    return bn_function
```

**Comment by ZhengRui** [comment]
@gpleiss , I played a lot with different implementations of densenet model these two days, the reason why it is so slow in multi-gpu cases during backward seems to be related to the number of checkpoints used in the model. e.g. in my 4 1080Ti desktop, with densenet201 using batch size 300, by default every denselayer will have a checkpoint, the backward will take around 36 seconds, but if you only add checkpoints after each denseblock, with only 4 checkpoints, the backward time decreased to be around 3-5 seconds. In the pytorch 2 old models, same batch size's backward takes 1.5s. I've also tested adding some checkpoints every n denselayers and i can see the decrease of backward time when n increase. However my modification of your model file seems to have some bug that cause gpu memory leak, but i believe the reason of this issue in your model is too many checkpoints after each denselayer. You can try to only add checkpoints after each denseblock, I did but has some bug mentioned before.

After all, @prigoyal 's model file https://github.com/prigoyal/pytorch_memonger/blob/master/models/optimized/densenet_new.py works really well. With same batch size(using chunks=8), the forward + backward total time is 3 seconds while previously the total time is 2 seconds which is quite acceptable. And it actually can support even slightly larger batch size as well. Maybe you can consider using this model file in your repo, I've compared forward and backward gradients, it is the same as your model.

**Comment by hendrycks** [comment]
This repository has faster checkpointing for multiple GPUs: https://github.com/csrhddlam/pytorch-checkpoint

**Comment by dbbert** [comment]
I experienced a ~3x slowdown during training when upgrading my checkpoint-heavy multi-gpu code from pytorch 0.4.1 to 1.0.1. I have not investigated the root cause, but reverting ```torch.utils.checkpoint``` to a previous version of the file solved the issue for me. So the problem must be in one of the latest commits.

**Comment by ferrine** [comment]
@dbbert, my guess is that bottleneck you observe might be due to this function
https://github.com/pytorch/pytorch/blob/master/torch/random.py#L48
As said in description this is slow for multigpu. Check your `preserve_rng_state` is set to False and you have deterministic output.

**Comment by yf225** [comment]
@dbbert Could you provide us with the minimal reproduction code for the slowdown issue?

**Comment by codertimo** [comment]
Is there any updates on this issue?
